text/3519-arbitrary-self-types-v2.md

```diff
@@ -228,15 +228,46 @@ where
 
 (See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
 
-It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+It is also implemented for `&T`, `&mut T`, `Weak<T>`, `NonNull<T>`, `*const T` and `*mut T`.
 
-## Compiler changes
+## Compiler changes: method probing
 
-The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html), and there's more detail in the [rustc dev guide](https://rustc-dev-guide.rust-lang.org/method-lookup.html).
 
-(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
+The key part of the first page is this:
 
-Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+> Then, for each candidate type `T`, search for a visible method with a receiver of that type in the following places:
+> - `T`'s inherent methods (methods implemented directly on `T`).
+> Any of the methods provided by a visible trait implemented by `T`.
```

Comment:
Nit: I think it makes more sense to quote another part of the reference, specifically a paragraph a bit higher up, the  one that starts with "The first step is to build a list of candidate receiver types. Obtain these by [...]"

Reply:
Done, thanks